### PR TITLE
Rover: sending of the new mavlink "WHEEL_DISTANCE" message added.

### PR DESCRIPTION
**Pull request by pavloblindnology**
Corresponding PR is here [ardupilot/mavlink #55](https://github.com/ArduPilot/mavlink/pull/55).

### CODE REVIEW

File: APMrover2/Rover.h [hunk 1]
```diff
@@ -368,6 +368,7 @@ class Rover : public AP_HAL::HAL::Callbacks {
 
     // last wheel encoder update times
     float wheel_encoder_last_angle_rad[WHEELENCODER_MAX_INSTANCES];     // distance in radians at time of last update to EKF
+    float wheel_encoder_last_distance_m[4];                             // distance in meters at time of last update to EKF (for reporting to GCS, use length 4 to fit the mavlink message)
```
Comment (khancyr) [inline APMrover2/Rover.h]: why 4 ? it shouldn't be WHEELENCODER_MAX_INSTANCES ?

File: libraries/AP_WheelEncoder/AP_WheelEncoder.cpp [hunk 2]
```diff
@@ -227,12 +227,13 @@ float AP_WheelEncoder::get_wheel_radius(uint8_t instance) const
...
-// get the total distance travelled in meters
-Vector3f AP_WheelEncoder::get_position(uint8_t instance) const
+// return a 3D vector defining the position offset of the center of the wheel in meters relative to the body frame origin
+const Vector3f &AP_WheelEncoder::get_pos_offset(uint8_t instance) const
 {
+    static Vector3f zeroVec;
     // for invalid instances return zero vector
     if (instance >= WHEELENCODER_MAX_INSTANCES) {
-        return Vector3f();
+        return zeroVec;
```
Comment (khancyr) [inline libraries/AP_WheelEncoder/AP_WheelEncoder.cpp]: If I follow correctly you use a static Vector3 that is zero to prevent create a zero vector. I think it is not worthly as the case where it is needed aren't numerous, no ?

File: APMrover2/sensors.cpp [hunk 3]
```diff
@@ -128,14 +128,17 @@ void Rover::update_wheel_encoder()
...
         wheel_encoder_last_angle_rad[i] = curr_angle_rad;
 
+        // save cumulative distances at current time (in meters)
+        wheel_encoder_last_distance_m[i] = g2.wheel_encoder.get_distance(i);
+
         // calculate delta time
         float delta_time;
         const uint32_t latest_sensor_update_ms = g2.wheel_encoder.get_last_reading_ms(i);
-        const uint32_t sensor_diff_ms = latest_sensor_update_ms - wheel_encoder_last_update_ms[i];
+        const int32_t sensor_diff_ms = latest_sensor_update_ms - wheel_encoder_last_update_ms[i];
```
Comment (khancyr) [inline APMrover2/sensors.cpp]: why this change ? normally we shouldn't have negative value

File: APMrover2/Rover.h (hunk 1 above)
Comment (pavloblindnology) [reply APMrover2/Rover.h]: The corresponding WHEEL_DISTANCE message holds array of size 4. So I use wheel_encoder_last_distance_m[4] to pass it to mavlink_msg_wheel_distance_send() procedure, which needs size 4 array.

File: libraries/AP_WheelEncoder/AP_WheelEncoder.cpp (hunk 2 above)
Comment (pavloblindnology) [reply libraries/AP_WheelEncoder/AP_WheelEncoder.cpp]: I use static vector because get_pos_offset() function returns by reference.

File: APMrover2/sensors.cpp (hunk 3 above)
Comment (pavloblindnology) [reply APMrover2/sensors.cpp]: But we have - for example, when wheel is not moving. Then wheel_encoder.get_last_reading_ms(i) reports some old time stamp, while wheel_encoder_last_update_ms[i] is always updated with the last EKF update (wheel_encoder_last_ekf_update_ms).

**Comment by pavloblindnology** [comment]
Any updates on this?

File: APMrover2/Rover.h (hunk 1 above)
Comment (AlexisTM) [reply APMrover2/Rover.h]: Probably because we might have up to 4 driven wheels?

**Comment by pavloblindnology** [comment]
@mikerob `RPM`-s will be available independently of `WHEEL_DISTANCE`-s.
The structure doesn't carry time delta, it carries timestamp. So you can calc the difference between two measurements and this gives you the delta. I have to note that currently there is no time synchronization between companion computer and Pixhawk - so that those timestamps are actually ignored - this results in noisy variable time deltas and speeds that only resembles the actual ones. So, using `WHEEL_DISTANCE`-s is better for total wheel odometry while `RPM`-s for speed computations.

**Comment by pavloblindnology** [comment]
Updated according to the last update of the message definition (up to 16 wheels supported now).

**Comment by pavloblindnology** [comment]
@mikerob Mavros plugin has been updated to compute speeds from distances using internal time stamps. This gives much more accurate speeds.
Concerning including time deltas to the message itself - it's not good since some packets can be dropped in which case there is no use of those deltas.

**Comment by rmackay9** [comment]
@pavloblindnology, This looks like a somewhat important PR but I don't quite understand what it's for.  Does it allow ROS to do something with the wheel encoder data?

**Comment by pavloblindnology** [comment]
@rmackay9 Yes, on the ROS side I have plugin to mavros which computes wheel odometry/speeds (including yaw) either from this new message or from RPM message. Afterwards this odometry can be fed to some fusing ROS package (like robot_localization, etc). So, Pixhawk serves only as a wheel encoder information provider. Currently, wheel odometry in Ardupilot is rather raw - e.g. it doesn't support yaw rate computations. With this approach, I can use whatever wheel odometry algorithm I want.

**Comment by rmackay9** [comment]
We've got a new version of this PR here so closing this one. https://github.com/ArduPilot/ardupilot/pull/10349
